pkg/migration/migration.go

```diff
@@ -0,0 +1,90 @@
+package migration
+
+import (
+	"fmt"
+
+	"k8s.io/client-go/kubernetes"
+
+	"github.com/sapcc/kubernikus/pkg/apis/kubernikus/v1"
+	kubernikus "github.com/sapcc/kubernikus/pkg/generated/clientset"
+	kubernikusfake "github.com/sapcc/kubernikus/pkg/generated/clientset/fake"
+)
+
+var defaultRegistry Registry
+
+// Migration describes an individual migration step.
+// The klusterRaw and kluster contain the to be migrate kluster.
+// The function is expected to modify the kluster accordingly, changed object is persisted
+// automatically after the handler returns with no error.
+// The kubernetes client can be used to modify other things (e.g. kluster secret)
+type Migration func(klusterRaw []byte, kluster *v1.Kluster, client kubernetes.Interface) (err error)
+
+//Latest returns to latest spec version available
+func Latest() int {
+	return defaultRegistry.Latest()
+}
+
+//MigrationsPending returns true if a kluster needs to be migrated
+func MigrationsPending(kluster *v1.Kluster) bool {
+	return defaultRegistry.MigrationsPending(kluster)
+}
+
+//Migrate a kluster to the most recent spec version
+func Migrate(k *v1.Kluster, client kubernetes.Interface, kubernikus_client kubernikus.Interface) error {
+	return defaultRegistry.Migrate(k, client, kubernikus_client)
+}
+
+//Registry manages an ordered list of migration steps
+type Registry struct {
+	migrations []Migration
+}
+
+//AddMigration appends a migration to the list
+func (r *Registry) AddMigration(m Migration) {
```

Comment:
Hm. Coming from Rails where the migrations are ordered by a unique date prefix, I was confused how this works here. I think, this `Registry` assumes that the migrations are always added at the end of the `register` function. And that there is no change in order or deletions.

I think, it's cool to keep it simple. But we should probably add a comment into the register call to not change it. Especially to not delete previously commited migrations.

Reply:
You are absolutely right with your assessments. I'll add the comment and would not invest more time at this point to make it more like a full-blown migration system (e.g. as in rails).

Reply:
Cool!